File: data/mesh_ingestor/ingestors.py
# Copyright © 2025-26 l5yth & contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Helpers for tracking ingestor identity and liveness announcements."""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Callable

from .. import VERSION as INGESTOR_VERSION
from . import config, queue
from .serialization import _canonical_node_id

HEARTBEAT_INTERVAL_SECS = 60 * 60
"""Default interval between ingestor heartbeat announcements."""


@dataclass
class _IngestorState:
    """Mutable ingestor identity and heartbeat tracking data."""

    start_time: int = field(default_factory=lambda: int(time.time()))
    last_heartbeat: int | None = None
    node_id: str | None = None


STATE = _IngestorState()
"""Shared ingestor identity state."""


def ingestor_start_time() -> int:
    """Return the unix timestamp representing when the ingestor booted."""

    return STATE.start_time


def set_ingestor_node_id(node_id: str | None) -> str | None:
    """Record the canonical host node identifier for the ingestor.

    Parameters:
        node_id: Raw node identifier reported by the connected device.

    Returns:
        Canonical node identifier in ``!xxxxxxxx`` form or ``None`` when the
        provided value cannot be normalised.
    """

    canonical = _canonical_node_id(node_id)
    if canonical is None:
        return None

    if STATE.node_id != canonical:
        STATE.node_id = canonical
        STATE.last_heartbeat = None

    return canonical


def queue_ingestor_heartbeat(
    *,
    force: bool = False,
    send: Callable[[str, dict], None] | None = None,
    node_id: str | None = None,
) -> bool:
    """Queue a heartbeat payload advertising ingestor liveness.

    Parameters:
        force: When ``True``, bypasses the heartbeat interval guard so an
            announcement is queued immediately.
        send: Optional transport callable used for tests; defaults to the queue
            dispatcher.
        node_id: Optional node identifier to register before sending. When
            omitted the previously recorded identifier is reused.

    Returns:
        ``True`` when a heartbeat payload was queued, ``False`` otherwise.
    """

    canonical = _canonical_node_id(node_id) if node_id is not None else None
    if canonical:
        set_ingestor_node_id(canonical)
    canonical = STATE.node_id

    if canonical is None:
        return False

    now = int(time.time())
    interval = max(
        0, int(getattr(config, "_INGESTOR_HEARTBEAT_SECS", HEARTBEAT_INTERVAL_SECS))
    )
    last = STATE.last_heartbeat
    if not force and last is not None and now - last < interval:
        return False

    payload = {
        "node_id": canonical,
        "start_time": STATE.start_time,
        "last_seen_time": now,
        "version": INGESTOR_VERSION,
    }
    queue._queue_post_json(
        "/api/ingestors",
        payload,
        priority=getattr(
            queue, "_INGESTOR_POST_PRIORITY", queue._DEFAULT_POST_PRIORITY
        ),
        send=send,
    )
    STATE.last_heartbeat = now
    return True


__all__ = [
    "HEARTBEAT_INTERVAL_SECS",
    "STATE",
    "ingestor_start_time",
    "queue_ingestor_heartbeat",
    "set_ingestor_node_id",
]

File: web/lib/potato_mesh/application/data_processing/ingestors.rb
# Copyright © 2025-26 l5yth & contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# frozen_string_literal: true

module PotatoMesh
  module App
    module DataProcessing
      # Insert or update an ingestor heartbeat payload.
      #
      # @param db [SQLite3::Database] open database handle.
      # @param payload [Hash] ingestor payload from the collector.
      # @return [Boolean] true when persistence succeeded.
      def upsert_ingestor(db, payload)
        return false unless payload.is_a?(Hash)

        parts = canonical_node_parts(payload["node_id"] || payload["id"])
        return false unless parts

        node_id, = parts
        now = Time.now.to_i

        start_time = coerce_integer(payload["start_time"] || payload["startTime"]) || now
        last_seen_time =
          coerce_integer(payload["last_seen_time"] || payload["lastSeenTime"]) || start_time

        start_time = 0 if start_time.negative?
        last_seen_time = 0 if last_seen_time.negative?
        start_time = now if start_time > now
        last_seen_time = now if last_seen_time > now
        last_seen_time = start_time if last_seen_time < start_time

        version = string_or_nil(payload["version"] || payload["ingestorVersion"])
        return false unless version
        lora_freq = coerce_integer(payload["lora_freq"])
        modem_preset = string_or_nil(payload["modem_preset"])
        protocol = string_or_nil(payload["protocol"]) || "meshtastic"

        with_busy_retry do
          db.execute <<~SQL, [node_id, start_time, last_seen_time, version, lora_freq, modem_preset, protocol]
                       INSERT INTO ingestors(node_id, start_time, last_seen_time, version, lora_freq, modem_preset, protocol)
                            VALUES(?,?,?,?,?,?,?)
                       ON CONFLICT(node_id) DO UPDATE SET
                         start_time = CASE
                           WHEN excluded.start_time > ingestors.start_time THEN excluded.start_time
                           ELSE ingestors.start_time
                         END,
                         last_seen_time = CASE
                           WHEN excluded.last_seen_time > ingestors.last_seen_time THEN excluded.last_seen_time
                           ELSE ingestors.last_seen_time
                         END,
                         version = COALESCE(excluded.version, ingestors.version),
                         lora_freq = COALESCE(excluded.lora_freq, ingestors.lora_freq),
                         modem_preset = COALESCE(excluded.modem_preset, ingestors.modem_preset),
                         protocol = excluded.protocol
                     SQL
        end

        true
      rescue SQLite3::SQLException => e
        warn_log(
          "Failed to upsert ingestor record",
          context: "data_processing.ingestors",
          node_id: node_id,
          error_class: e.class.name,
          error_message: e.message,
        )
        false
      end
    end
  end
end
